src/com/photoncat/aiproj2/heuristic/Heuristic2.java

```diff
@@ -0,0 +1,902 @@
+    @Override
+    public int heuristic(Board board){
+        int n=board.getSize();
+        int m=board.getM();
+
+        if (n==3&&m==3){
+            return ThreeByThree(board);
+        }
+        if (m==5){
+            return NByFive(board);
+        }
+        if (m==4){
+            return NByFour(board);
+        }
+        else{
+            return NbyM(board);
+        }
+//        if ()
+//        return 0;
+    }
+
+    private int NbyM(Board board) {
+        if (checkWin(board,PieceType.CROSS)){
+            return 50000000;
+        }
+        else if (checkWin(board,PieceType.CIRCLE)){
+            return -50000000;
+        }
+        else {
+            int sum=0;
+            int N=board.getSize();
+            for(int i=0;i<N;i++){
+                for (int j=0;j<N;j++){
+                    if (board.getPiece(i,j)==PieceType.NONE){
+                        sum=sum+evaluateNByM(board,i,j,PieceType.CROSS);
+                        sum=sum-evaluateNByM(board,i,j,PieceType.CIRCLE);
+                    }
+
+                }
+            }
+            return sum;
+        }
+    }
+    private int NByFive(Board board) {
+        if (checkWin(board,PieceType.CROSS)){
+            return 10000000;
+        }
+        else if (checkWin(board,PieceType.CIRCLE)){
+            return -10000000;
+        }
+        else {
+            int sum=0;
+            int N=board.getSize();
+            for(int i=0;i<N;i++){
+                for (int j=0;j<N;j++){
+                    if (board.getPiece(i,j)==PieceType.NONE){
+//                         sum=sum+evaluateNByFive(board,i,j,PieceType.CROSS);
+//                         sum=sum-evaluateNByFive(board,i,j,PieceType.CIRCLE);
+//                         sum=sum+evaluateNByFive2(board,i,j,PieceType.CROSS);
+//                         sum=sum-evaluateNByFive2(board,i,j,PieceType.CIRCLE);
+//                         sum=sum+evaluateNByM(board,i,j,PieceType.CROSS);
+//                         sum=sum-evaluateNByM(board,i,j,PieceType.CIRCLE);
+                        sum=sum+evaluateNByFive3(board,i,j,PieceType.CROSS);
+                        sum=sum-evaluateNByFive3(board,i,j,PieceType.CIRCLE);
+                    }
+
+                }
+            }
+            return sum;
+        }
+    }
+    private int NByFour(Board board) {
+        if (checkWin(board,PieceType.CROSS)){
+            return 50000000;
+        }
+        else if (checkWin(board,PieceType.CIRCLE)){
+            return -50000000;
+        }
+        else {
+            int sum=0;
+            int N=board.getSize();
+            for(int i=0;i<N;i++){
+                for (int j=0;j<N;j++){
+                    if (board.getPiece(i,j)==PieceType.NONE){
+                        sum=sum+evaluateNByFour(board,i,j,PieceType.CROSS);
+                        sum=sum-evaluateNByFour(board,i,j,PieceType.CIRCLE);
+                    }
+
+                }
+            }
+            return sum;
+        }
+    }
```

Comment:
Above three functions are extremely alike. Please consider using a delegate instead of copying the same code several times.